### ISSUE TITLE
Performance regression

### ISSUE THREAD

**Issue by Jessime** [wweir827/CHAMP#7]
I haven't had a chance to look too much into this, but I just updated CHAMP for the first time since ~ March. Based on "Return time", it looks like `parallel_louvain` is running 2-4x slower than it was. Is there any explanation for this?

**Comment by wweir827** [comment]
I think this might be from the switch to the new louvain-igraph package ( which was drastically rewritten by traag.).  I'm probably going to look into switching to this package for generating the partitions : https://github.com/vtraag/leidenalg which is supposed to have better performance.  

Have to make sure it's compatible though.

**Comment by Jessime** [comment]
I saw this paper a couple months ago when Natalie posted it! I went ahead and did a couple simple tests on one of my graphs. Leidenalg is faster:

```python
In [20]: %timeit partition_louvain = louvain.find_partition(graph, louvain.RBConfigurationVertexPartition, weights
    ...: ='weight', resolution_parameter=2)                                                                       
24.6 s ± 9.12 s per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [21]: %timeit partition_leiden = leidenalg.find_partition(graph, leidenalg.RBConfigurationVertexPartition, weig
    ...: hts='weight', n_iterations=2, resolution_parameter=2)                                                    
11 s ± 1.8 s per loop (mean ± std. dev. of 7 runs, 1 loop each)

```
and ever so slightly better:

```python
In [25]: for i in range(10): 
    ...:     print(louvain.find_partition(graph, louvain.RBConfigurationVertexPartition, weights='weight', resolut
    ...: ion_parameter=2).modularity) 
    ...:                                                                                                          
0.6301019439537691
0.6333852361832993
0.6269146951348228
0.6373721188287049
0.6374519633844057
0.6291527723806505
0.6299580723991539
0.6289395297921448
0.6305391387766376
0.6378631596151206

In [24]: for i in range(10): 
    ...:     print(leidenalg.find_partition(graph, leidenalg.RBConfigurationVertexPartition, weights='weight', n_i
    ...: terations=2, resolution_parameter=2).modularity) 
    ...:                                                                                                          
0.6348178823876356
0.6361116165666902
0.6376424392900716
0.6305302906789212
0.6392127460349585
0.6392124048932231
0.637206816545722
0.640510432372692
0.6346067527392649
0.6349438918525236
```

Seems like an easy win to me.

**Comment by Jessime** [comment]
> Seems like an easy win to me.

Follow up. It seems like the communities called by Leiden are less consistent between runs. I don't know if this matters for CHAMP, but depending on the application, it could be considered a downside. It messed with some of the pipeline I've been building. 

I wanted to quantify the difference in stability, to make sure I wasn't just imagining it. I'm not sure this is the most rigorous way of doing it, but I called communities 10 times, did pairwise comparisons between calls using `sklearn.metrics.normalized_mutual_info_score`, and collected the results:

**Louvain:**

```python
DescribeResult(nobs=45, minmax=(0.739, 0.902), mean=0.821, variance=0.00118, skewness=-0.185, kurtosis=0.039)
```

and **Leiden:**

```python
DescribeResult(nobs=45, minmax=(0.636, 0.762), mean=0.7104, variance=0.00066, skewness=-0.284, kurtosis=0.231)
```

So, this might be another consideration.

**Comment by wweir827** [comment]
Sorry been a little while since i've had time to think about this.  In traag's paper on this, they suggest running Leiden iteratively (with output of previous run as the input of next run).  It's possible that this will give it more consistency from run to run.  There is also an additional bit of the algorithm that is stochastic ( whether or not it merges communities during the refinement step) which could also be causing more variability.  I think what i'm going to do is just add a method to call leiden in the same way in parallel for the single layer case, with an option to set the number of back to back iterative calls at each run.

**Comment by vtraag** [comment]
@jessime that might be something to look into further. Typically the Leiden algorithm will find smaller communities, which were incorrectly merged together by the Louvain algorithm. My bet is that the larger variability is mostly due to this. 

Running for multiple iterations should presumably reduce the variability over multiple runs somewhat, as @wweir827  suggest. You can simply specify `n_iterations`, the default is `n_iterations=2`. 

Depending on your application needs, you may also simply specify a [seed](https://leidenalg.readthedocs.io/en/latest/reference.html#leidenalg.Optimiser.set_rng_seed) for the random number generator to ensure the outcome is always replicable.
